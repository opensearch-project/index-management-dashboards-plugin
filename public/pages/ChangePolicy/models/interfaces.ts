/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Policy } from "../../../../models/interfaces";

export interface PolicyOption {
  label: string;
  value: Policy;
}
