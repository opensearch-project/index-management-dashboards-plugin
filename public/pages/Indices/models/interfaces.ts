/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";
import { Policy } from "../../../../models/interfaces";
import { ManagedCatIndex } from "../../../../server/models/interfaces";

export interface PolicyOption {
  label: string;
  policy?: Policy;
}

export interface IndicesQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof ManagedCatIndex;
  sortDirection: Direction;
  showDataStreams: boolean;
}
