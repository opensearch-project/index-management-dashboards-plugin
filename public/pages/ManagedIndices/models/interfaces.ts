/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";
import { ManagedIndexItem } from "../../../../models/interfaces";

export interface ManagedIndicesQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof ManagedIndexItem;
  sortDirection: Direction;
  showDataStreams: boolean;
}
