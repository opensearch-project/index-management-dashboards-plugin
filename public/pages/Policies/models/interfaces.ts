/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";

export interface PolicyItem {
  id: string;
  seqNo: number;
  primaryTerm: number;
  policy: object; // only dumped to view as JSON as of now, don't need to type
}

export interface PoliciesQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof PolicyItem;
  sortDirection: Direction;
}
