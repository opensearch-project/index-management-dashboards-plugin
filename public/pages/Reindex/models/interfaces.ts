/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Conflicts } from "elasticsearch";

export interface ReindexRequest {
  waitForCompletion: boolean;
  slices?: number | string;
  maxDocs?: number;
  body: {
    conflicts?: Conflicts;
    source: {
      index: string;
      [key: string]: any;
    };
    dest: {
      index: string;
      pipeline?: string;
      op_type?: string;
    };
  };
  script?: {
    source: string;
    lang: string;
  };
}
export interface ReindexResponse {
  task: string;
}

export interface IndexSelectItem {
  status?: string;
  health?: string;
  isIndex?: boolean;
  isDataStream?: boolean;
  isAlias?: boolean;
  indices?: string[];
  writingIndex?: string;
}
