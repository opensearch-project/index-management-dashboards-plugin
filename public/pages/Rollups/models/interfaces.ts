/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";
import { DocumentRollup } from "../../../../models/interfaces";

export interface RollupQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof DocumentRollup;
  sortDirection: Direction;
}
