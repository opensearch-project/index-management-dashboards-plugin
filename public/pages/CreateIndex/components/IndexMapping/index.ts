/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import IndexMapping from "./IndexMapping";

export * from "./IndexMapping";
export default IndexMapping;
