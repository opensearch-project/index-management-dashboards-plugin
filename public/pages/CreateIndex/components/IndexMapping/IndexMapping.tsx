/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { forwardRef, useCallback, useState, Ref, useRef, useMemo, useImperativeHandle } from "react";
import { EuiTreeView, EuiIcon, EuiTreeViewProps, EuiButton, EuiSpacer, EuiButtonGroup, EuiLink } from "@elastic/eui";
import { set, get, isEmpty } from "lodash";
import JSONEditor, { IJSONEditorRef } from "../../../../components/JSONEditor";
import { Modal } from "../../../../components/Modal";
import { MappingsProperties } from "../../../../../models/interfaces";
import CustomFormRow from "../../../../components/CustomFormRow";
import MappingLabel, { IMappingLabelRef } from "../MappingLabel";
import { transformObjectToArray, transformArrayToObject, countNodesInTree } from "./helper";
import { IndexMappingsObjectAll, IndexMappingProps, EDITOR_MODE, IIndexMappingsRef } from "./interfaces";
import "./IndexMapping.scss";

export * from "./helper";
export * from "./interfaces";

const IndexMapping = (
  { value: propsValue, onChange: propsOnChange, isEdit, oldValue, readonly, docVersion }: IndexMappingProps,
  ref: Ref<IIndexMappingsRef>
) => {
  const value = propsValue?.properties || [];
  const onChange = (val: MappingsProperties) => {
    propsOnChange({
      ...propsValue,
      properties: val,
    });
  };
  const allFieldsRef = useRef<Record<string, IMappingLabelRef>>({});
  const JSONEditorRef = useRef<IJSONEditorRef>(null);
  useImperativeHandle(ref, () => ({
    validate: async () => {
      const values = await Promise.all(Object.values(allFieldsRef.current).map((item) => item.validate()));
      const JSONEditorValidateResult = await JSONEditorRef.current?.validate();
      return values.some((item) => item) || JSONEditorValidateResult ? "with error" : "";
    },
    getJSONEditorValue: () => JSONEditorRef.current?.getValue() || "",
  }));
  const [editorMode, setEditorMode] = useState<EDITOR_MODE>(EDITOR_MODE.VISUAL);
  const addField = useCallback(
    (pos, fieldSettings?: Partial<MappingsProperties[number]>) => {
      const newValue = [...(value || [])];
      const nowProperties = ((pos ? get(newValue, pos) : (newValue as MappingsProperties)) || []) as MappingsProperties;
      nowProperties.push({
        fieldName: fieldSettings?.fieldName || "",
        type: "text",
        ...fieldSettings,
      });
      if (pos) {
        set(newValue, pos, nowProperties);
      }
      onChange(newValue);
    },
    [onChange, value]
  );
  const deleteField = useCallback(
    (pos) => {
      const newValue = [...(value || [])];
      const splittedArray = pos.split(".");
      const index = splittedArray[splittedArray.length - 1];
      const prefix = splittedArray.slice(0, -1);
      const prefixPos = prefix.join(".");
      const nowProperties = ((prefixPos ? get(newValue, prefixPos) : (newValue as MappingsProperties)) || []) as MappingsProperties;
      nowProperties.splice(index, 1);

      if (prefixPos) {
        set(newValue, prefixPos, nowProperties);
      }

      onChange(newValue);
    },
    [onChange, value]
  );
  const transformValueToTreeItems = (formValue: MappingsProperties, pos: string = ""): EuiTreeViewProps["items"] => {
    let isFirstEditableField = false;
    return (formValue || []).map((item, index) => {
      const { fieldName, ...fieldSettings } = item;
      const id = [pos, index].filter((item) => item !== "").join(".properties.");
      const readonlyFlag = readonly || (isEdit && !!get(oldValue?.properties, id));
      let shouldShowLabel = false;
      if (!readonlyFlag && !isFirstEditableField) {
        isFirstEditableField = true;
        shouldShowLabel = true;
      }
      const payload: EuiTreeViewProps["items"][number] = {
        label: (
          <MappingLabel
            shouldShowLabel={shouldShowLabel}
            ref={(ref) => {
              if (ref) {
                allFieldsRef.current[id] = ref;
              } else {
                delete allFieldsRef.current[id];
              }
            }}
            readonly={readonlyFlag}
            value={item}
            id={`mapping-visual-editor-${id}`}
            onFieldNameCheck={(fieldName) => {
              const hasDuplicateName = (formValue || [])
                .filter((sibItem, sibIndex) => sibIndex < index)
                .some((sibItem) => sibItem.fieldName === fieldName);
              if (hasDuplicateName) {
                return `Duplicate field name [${fieldName}], please change your field name`;
              }

              return "";
            }}
            onChange={(val, key, v) => {
              const newValue = [...(value || [])];
              set(newValue, id, val);
              onChange(newValue);
            }}
            onDeleteField={() => {
              deleteField(id);
            }}
            onAddSubField={() => {
              addField(`${id}.properties`);
            }}
            onAddSubObject={() => {
              addField(`${id}.properties`, {
                type: "object",
              });
            }}
          />
        ),
        id: `mapping-visual-editor-${id}`,
        icon: <EuiIcon type="arrowRight" style={{ visibility: "hidden" }} />,
        iconWhenExpanded: <EuiIcon type="arrowDown" style={{ visibility: "hidden" }} />,
      };
      if (fieldSettings.properties) {
        (payload.icon = <EuiIcon type="arrowRight" />),
          (payload.iconWhenExpanded = <EuiIcon type="arrowDown" />),
          (payload.children = transformValueToTreeItems(fieldSettings.properties, id));
      }

      return payload;
    });
  };
  const transformedTreeItems = useMemo(() => transformValueToTreeItems(value), [value]);
  const newValue = useMemo(() => {
    const oldValueKeys = (oldValue?.properties || []).map((item) => item.fieldName);
    return value?.filter((item) => !oldValueKeys.includes(item.fieldName)) || [];
  }, [oldValue?.properties, value]);
  const renderKey = useMemo(() => {
    return countNodesInTree(value || []);
  }, [value]);
  return (
    <>
      <EuiButtonGroup
        type="single"
        idSelected={editorMode as string}
        onChange={(id) => setEditorMode(id as EDITOR_MODE)}
        legend="Editor Type"
        options={[
          {
            label: readonly ? "Tree view" : "Visual Editor",
            id: EDITOR_MODE.VISUAL,
            "data-test-subj": "editorTypeVisualEditor",
          },
          {
            label: readonly ? "JSON" : "JSON Editor",
            id: EDITOR_MODE.JSON,
            "data-test-subj": "editorTypeJsonEditor",
          },
        ]}
      />
      <EuiSpacer />
      {editorMode === EDITOR_MODE.VISUAL ? (
        <>
          {transformedTreeItems.length ? (
            <EuiTreeView
              key={renderKey}
              expandByDefault={!readonly}
              className="index-mapping-tree"
              aria-labelledby="label"
              items={transformValueToTreeItems(value)}
            />
          ) : (
            <p>You have no field mappings.</p>
          )}
          {readonly ? null : (
            <>
              <EuiSpacer />
              <EuiButton style={{ marginRight: 8 }} data-test-subj="createIndexAddFieldButton" onClick={() => addField("")}>
                Add new field
              </EuiButton>
              <EuiButton
                data-test-subj="createIndexAddObjectFieldButton"
                onClick={() =>
                  addField("", {
                    type: "object",
                  })
                }
              >
                Add new object
              </EuiButton>
            </>
          )}
        </>
      ) : (
        <>
          {isEdit && !readonly && !isEmpty(oldValue) ? (
            <>
              <EuiButton
                size="s"
                data-test-subj="previousMappingsJsonButton"
                onClick={() => {
                  Modal.show({
                    title: "Previous mappings",
                    content: (
                      <JSONEditor
                        readOnly
                        value={JSON.stringify(
                          {
                            ...oldValue,
                            properties: transformArrayToObject(oldValue?.properties || []),
                          },
                          null,
                          2
                        )}
                      />
                    ),
                    "data-test-subj": "previousMappingsJsonModal",
                    onOk: () => {},
                  });
                }}
              >
                See previous mappings
              </EuiButton>
              <EuiSpacer />
            </>
          ) : null}
          {readonly ? (
            <JSONEditor
              ref={JSONEditorRef}
              value={JSON.stringify(
                {
                  ...propsValue,
                  properties: transformArrayToObject(value || []),
                },
                null,
                2
              )}
              readOnly={readonly}
              width="100%"
            />
          ) : (
            <CustomFormRow
              data-test-subj="mappingsJsonEditorFormRow"
              label="Specify index mapping"
              helpText={
                <>
                  Specify mapping in JSON format.{" "}
                  <EuiLink
                    external
                    target="_blank"
                    href={`https://opensearch.org/docs/${docVersion}/opensearch/mappings/#mapping-example-usage`}
                  >
                    View mapping example.
                  </EuiLink>
                </>
              }
              fullWidth
            >
              <JSONEditor
                value={JSON.stringify(
                  {
                    ...propsValue,
                    properties: transformArrayToObject(newValue || []),
                  },
                  null,
                  2
                )}
                onChange={(val) => {
                  const result: IndexMappingsObjectAll = JSON.parse(val);
                  propsOnChange({
                    ...result,
                    properties: [...(oldValue?.properties || []), ...transformObjectToArray(result?.properties || {})],
                  });
                }}
                width="100%"
                ref={JSONEditorRef}
              />
            </CustomFormRow>
          )}
        </>
      )}
    </>
  );
};

export default forwardRef(IndexMapping);
