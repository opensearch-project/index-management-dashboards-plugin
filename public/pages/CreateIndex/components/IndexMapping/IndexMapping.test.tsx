/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { forwardRef, Ref, useRef, useState } from "react";
import { render, fireEvent, waitFor, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import IndexMapping, { IIndexMappingsRef, IndexMappingProps, transformObjectToArray } from "./IndexMapping";
import { MappingsProperties } from "../../../../../models/interfaces";
import { renderHook } from "@testing-library/react-hooks";

const IndexMappingOnChangeWrapper = forwardRef((props: Partial<IndexMappingProps>, ref: Ref<IIndexMappingsRef>) => {
  const [value, setValue] = useState(props.value as any);
  return (
    <IndexMapping
      {...props}
      ref={ref}
      value={value}
      onChange={(val) => {
        setValue(val);
      }}
    />
  );
});

describe("<IndexMapping /> spec", () => {
  it("renders the component", () => {
    const { container } = render(<IndexMapping onChange={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("render mappings with object type", () => {
    const { container } = render(
      <IndexMapping
        onChange={() => {}}
        value={[{ fieldName: "object", type: "object", properties: [{ fieldName: "text", type: "text" }] }]}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("render mappings with oldValue in edit mode and all operation works well", async () => {
    const { result } = renderHook(() => {
      const ref = useRef<IIndexMappingsRef>(null);
      const renderResult = render(
        <IndexMappingOnChangeWrapper
          ref={ref}
          isEdit
          oldValue={[{ fieldName: "object", type: "object", properties: [{ fieldName: "text", type: "text" }] }]}
          value={[{ fieldName: "object", type: "object", properties: [{ fieldName: "text", type: "text" }] }]}
        />
      );

      return {
        renderResult,
        ref,
      };
    });
    const { renderResult, ref } = result.current;
    const { getByTestId, getByText, queryByTestId, queryByText } = renderResult;

    // old field disable check
    expect(getByTestId("mapping-visual-editor-0-field-name")).toHaveAttribute("disabled");
    expect(getByTestId("mapping-visual-editor-0.properties.0-field-name")).toHaveAttribute("disabled");
    expect(document.querySelector('[data-test-subj="mapping-visual-editor-0-delete-field"]')).toBeNull();
    expect(document.querySelector('[data-test-subj="mapping-visual-editor-0.properties.0-add-sub-field"]')).toBeNull();
    expect(document.querySelector('[data-test-subj="mapping-visual-editor-0.properties.0-delete-field"]')).toBeNull();

    // add a new field
    userEvent.click(getByTestId("create index add field button"));
    // new field should be editable
    expect(getByTestId("mapping-visual-editor-1-field-name")).not.toHaveAttribute("disabled");
    expect(document.querySelector('[data-test-subj="mapping-visual-editor-1-delete-field"]')).not.toBeNull();

    // empty and duplicate validation for field name
    userEvent.click(document.querySelector('[data-test-subj="mapping-visual-editor-1-field-name"]') as Element);
    expect(getByTestId("mapping-visual-editor-1-field-name")).toHaveValue("");
    expect(queryByText("Field name is required, please input")).toBeNull();
    userEvent.click(document.body);
    expect(getByText("Field name is required, please input")).not.toBeNull();
    userEvent.type(getByTestId("mapping-visual-editor-1-field-name"), "object");
    userEvent.click(document.body);
    expect(getByText("Duplicate field name [object], please change your field name")).not.toBeNull();
    await act(async () => {
      expect(await ref.current?.validate()).toEqual("with error");
    });
    userEvent.clear(getByTestId("mapping-visual-editor-1-field-name"));
    userEvent.type(getByTestId("mapping-visual-editor-1-field-name"), "new_object");
    await act(async () => {
      expect(await ref.current?.validate()).toEqual("");
    });
    expect(queryByText("Duplicate field name [object], please change your field name")).toBeNull();
    userEvent.click(document.body);

    // only show the sub action for type of object
    expect(queryByTestId("mapping-visual-editor-1-add-sub-field")).toBeNull();

    // change type to object
    fireEvent.change(getByTestId("mapping-visual-editor-1-field-type"), {
      target: {
        value: "object",
      },
    });

    // sub action for object
    expect(getByTestId("mapping-visual-editor-1-add-sub-field")).not.toBeNull();
    userEvent.click(getByTestId("mapping-visual-editor-1-add-sub-field"));
    // new sub field check
    expect((getByTestId("mapping-visual-editor-1.properties.0-field-type") as HTMLSelectElement).value).toBe("text");
    await waitFor(() => {
      userEvent.click(getByTestId("mapping-visual-editor-1.properties.0-delete-field"));
    });

    // add a new field
    userEvent.click(getByTestId("create index add field button"));
    // delete the new field
    await waitFor(() => {});
    userEvent.click(getByTestId("mapping-visual-editor-2-delete-field"));
    expect(queryByTestId("mapping-visual-editor-2-delete-field")).toBeNull();

    await userEvent.click(getByTestId("editor-type-json-editor").querySelector("input") as Element);
    await waitFor(() => {});
    userEvent.click(getByTestId("previous-mappings-json-button"));
    await waitFor(() => {});
    expect(queryByTestId("previous-mappings-json-modal-ok")).not.toBeNull();
    userEvent.click(getByTestId("previous-mappings-json-modal-ok"));
    await waitFor(() => {
      expect(queryByTestId("previous-mappings-json-modal-ok")).toBeNull();
    });
  });

  it("it transformObjectToArray", () => {
    expect(
      transformObjectToArray({
        test: {
          type: "text",
          properties: {
            test_children: {
              type: "text",
            },
          },
        },
      })
    ).toEqual([
      {
        fieldName: "test",
        type: "text",
        properties: [
          {
            fieldName: "test_children",
            type: "text",
          },
        ],
      },
    ] as MappingsProperties);
  });
});
