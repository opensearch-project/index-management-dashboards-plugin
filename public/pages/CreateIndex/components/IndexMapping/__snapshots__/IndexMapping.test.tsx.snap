// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IndexMapping /> spec render mappings with object type 1`] = `
<div>
  <fieldset
    class="euiButtonGroup euiButtonGroup--small euiButtonGroup--text"
  >
    <legend
      class="euiScreenReaderOnly"
    >
      Editor Type
    </legend>
    <div
      class="euiButtonGroup__buttons"
    >
      <label
        class="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small euiButtonGroupButton--fill euiButtonGroupButton-isSelected"
        data-test-subj="editor-type-visual-editor"
        for="some_html_id"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text euiButtonGroupButton__textShift"
            data-text="Visual Editor"
            title="Visual Editor"
          >
            <input
              checked=""
              class="euiScreenReaderOnly"
              data-test-subj="VISUAL"
              id="some_html_id"
              name="some_html_id"
              type="radio"
              value=""
            />
            Visual Editor
          </span>
        </span>
      </label>
      <label
        class="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small"
        data-test-subj="editor-type-json-editor"
        for="some_html_id"
      >
        <span
          class="euiButtonContent euiButton__content"
        >
          <span
            class="euiButton__text euiButtonGroupButton__textShift"
            data-text="JSON Editor"
            title="JSON Editor"
          >
            <input
              class="euiScreenReaderOnly"
              data-test-subj="JSON"
              id="some_html_id"
              name="some_html_id"
              type="radio"
              value=""
            />
            JSON Editor
          </span>
        </span>
      </label>
    </div>
  </fieldset>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div
    class="euiText euiText--medium euiTreeView__wrapper"
  >
    <p
      class="euiScreenReaderOnly"
      id="some_html_id--instruction"
    >
      You can quickly navigate this list using arrow keys.
    </p>
    <ul
      aria-describedby="some_html_id--instruction"
      aria-labelledby="label"
      class="euiTreeView index-mapping-tree"
      id="some_html_id"
    >
      <li
        class="euiTreeView__node euiTreeView__node--expanded"
      >
        <button
          aria-controls="some_html_id"
          aria-expanded="true"
          class="euiTreeView__nodeInner"
          data-test-subj="euiTreeViewButton-some_html_id"
          id="mapping-visual-editor-0"
        >
          <span
            class="euiTreeView__iconWrapper"
          >
            EuiIconMock
          </span>
          <span
            class="euiTreeView__nodeLabel"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
                style="width: 240px;"
              >
                <div
                  class="euiFormRow euiFormRow--compressed"
                  id="some_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="some_html_id"
                    >
                      Field name
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <span
                      class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                    >
                      <div
                        class="euiFormControlLayout euiFormControlLayout--compressed"
                      >
                        <div
                          class="euiFormControlLayout__childrenWrapper"
                        >
                          <input
                            class="euiFieldText euiFieldText--compressed"
                            data-test-subj="mapping-visual-editor-0-field-name"
                            disabled=""
                            id="some_html_id"
                            type="text"
                            value="object"
                          />
                        </div>
                      </div>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiFormRow euiFormRow--compressed"
                  id="some_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      class="euiFormLabel euiFormRow__label"
                      for="some_html_id"
                    >
                      Field type
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <span
                      class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                    >
                      <div
                        class="euiFormControlLayout euiFormControlLayout--compressed"
                      >
                        <div
                          class="euiFormControlLayout__childrenWrapper"
                        >
                          <select
                            class="euiSelect euiSelect--compressed"
                            data-test-subj="mapping-visual-editor-0-field-type"
                            disabled=""
                            id="some_html_id"
                          >
                            <option
                              value="null"
                            >
                              null
                            </option>
                            <option
                              value="boolean"
                            >
                              boolean
                            </option>
                            <option
                              value="float"
                            >
                              float
                            </option>
                            <option
                              value="double"
                            >
                              double
                            </option>
                            <option
                              value="integer"
                            >
                              integer
                            </option>
                            <option
                              value="text"
                            >
                              text
                            </option>
                            <option
                              value="keyword"
                            >
                              keyword
                            </option>
                            <option
                              value="date"
                            >
                              date
                            </option>
                            <option
                              value="object"
                            >
                              object
                            </option>
                          </select>
                          <div
                            class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                          >
                            <span
                              class="euiFormControlLayoutCustomIcon"
                            >
                              EuiIconMock
                            </span>
                          </div>
                        </div>
                      </div>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiFormRow euiFormRow--compressed"
                  id="some_html_id-row"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      class="euiFormLabel euiFormRow__label"
                      for="some_html_id"
                    >
                      actions
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      id="some_html_id"
                      style="display: flex; height: 32px; align-items: center;"
                    >
                      <span
                        class="euiToolTipAnchor"
                      >
                        <span
                          aria-label="Delete current field"
                          class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--medium"
                          data-test-subj="mapping-visual-editor-0-add-sub-field"
                        >
                          EuiIconMock
                        </span>
                      </span>
                      <span
                        class="euiToolTipAnchor"
                      >
                        <span
                          aria-label="Delete current field"
                          class="euiButtonIcon euiButtonIcon--danger euiButtonIcon--empty euiButtonIcon--medium"
                          data-test-subj="mapping-visual-editor-0-delete-field"
                        >
                          EuiIconMock
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </span>
        </button>
        <div
          id="some_html_id"
        >
          <div
            class="euiText euiText--medium euiTreeView__wrapper"
          >
            <ul
              aria-labelledby="mapping-visual-editor-0 label"
              class="euiTreeView"
            >
              <li
                class="euiTreeView__node"
              >
                <button
                  aria-controls="some_html_id"
                  aria-expanded="false"
                  class="euiTreeView__nodeInner"
                  data-test-subj="euiTreeViewButton-some_html_id"
                  id="mapping-visual-editor-0.properties.0"
                >
                  <span
                    class="euiTreeView__iconWrapper"
                  >
                    EuiIconMock
                  </span>
                  <span
                    class="euiTreeView__nodeLabel"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                        style="width: 240px;"
                      >
                        <div
                          class="euiFormRow euiFormRow--compressed"
                          id="some_html_id-row"
                        >
                          <div
                            class="euiFormRow__labelWrapper"
                          >
                            <label
                              aria-invalid="false"
                              class="euiFormLabel euiFormRow__label"
                              for="some_html_id"
                            >
                              Field name
                            </label>
                          </div>
                          <div
                            class="euiFormRow__fieldWrapper"
                          >
                            <span
                              class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                            >
                              <div
                                class="euiFormControlLayout euiFormControlLayout--compressed"
                              >
                                <div
                                  class="euiFormControlLayout__childrenWrapper"
                                >
                                  <input
                                    class="euiFieldText euiFieldText--compressed"
                                    data-test-subj="mapping-visual-editor-0.properties.0-field-name"
                                    disabled=""
                                    id="some_html_id"
                                    type="text"
                                    value="text"
                                  />
                                </div>
                              </div>
                            </span>
                          </div>
                        </div>
                      </div>
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <div
                          class="euiFormRow euiFormRow--compressed"
                          id="some_html_id-row"
                        >
                          <div
                            class="euiFormRow__labelWrapper"
                          >
                            <label
                              class="euiFormLabel euiFormRow__label"
                              for="some_html_id"
                            >
                              Field type
                            </label>
                          </div>
                          <div
                            class="euiFormRow__fieldWrapper"
                          >
                            <span
                              class="euiToolTipAnchor euiToolTipAnchor--displayBlock"
                            >
                              <div
                                class="euiFormControlLayout euiFormControlLayout--compressed"
                              >
                                <div
                                  class="euiFormControlLayout__childrenWrapper"
                                >
                                  <select
                                    class="euiSelect euiSelect--compressed"
                                    data-test-subj="mapping-visual-editor-0.properties.0-field-type"
                                    disabled=""
                                    id="some_html_id"
                                  >
                                    <option
                                      value="null"
                                    >
                                      null
                                    </option>
                                    <option
                                      value="boolean"
                                    >
                                      boolean
                                    </option>
                                    <option
                                      value="float"
                                    >
                                      float
                                    </option>
                                    <option
                                      value="double"
                                    >
                                      double
                                    </option>
                                    <option
                                      value="integer"
                                    >
                                      integer
                                    </option>
                                    <option
                                      value="text"
                                    >
                                      text
                                    </option>
                                    <option
                                      value="keyword"
                                    >
                                      keyword
                                    </option>
                                    <option
                                      value="date"
                                    >
                                      date
                                    </option>
                                    <option
                                      value="object"
                                    >
                                      object
                                    </option>
                                  </select>
                                  <div
                                    class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                                  >
                                    <span
                                      class="euiFormControlLayoutCustomIcon"
                                    >
                                      EuiIconMock
                                    </span>
                                  </div>
                                </div>
                              </div>
                            </span>
                          </div>
                        </div>
                      </div>
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <div
                          class="euiFormRow euiFormRow--compressed"
                          id="some_html_id-row"
                        >
                          <div
                            class="euiFormRow__labelWrapper"
                          >
                            <label
                              class="euiFormLabel euiFormRow__label"
                              for="some_html_id"
                            >
                              actions
                            </label>
                          </div>
                          <div
                            class="euiFormRow__fieldWrapper"
                          >
                            <div
                              id="some_html_id"
                              style="display: flex; height: 32px; align-items: center;"
                            >
                              <span
                                class="euiToolTipAnchor"
                              >
                                <span
                                  aria-label="Delete current field"
                                  class="euiButtonIcon euiButtonIcon--danger euiButtonIcon--empty euiButtonIcon--medium"
                                  data-test-subj="mapping-visual-editor-0.properties.0-delete-field"
                                >
                                  EuiIconMock
                                </span>
                              </span>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </span>
                </button>
                <div
                  id="some_html_id"
                />
              </li>
            </ul>
          </div>
        </div>
      </li>
    </ul>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <button
    class="euiButton euiButton--primary"
    data-test-subj="create index add field button"
    type="button"
  >
    <span
      class="euiButtonContent euiButton__content"
    >
      <span
        class="euiButton__text"
      >
        Add a field
      </span>
    </span>
  </button>
</div>
`;

exports[`<IndexMapping /> spec renders the component 1`] = `
<fieldset
  class="euiButtonGroup euiButtonGroup--small euiButtonGroup--text"
>
  <legend
    class="euiScreenReaderOnly"
  >
    Editor Type
  </legend>
  <div
    class="euiButtonGroup__buttons"
  >
    <label
      class="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small euiButtonGroupButton--fill euiButtonGroupButton-isSelected"
      data-test-subj="editor-type-visual-editor"
      for="some_html_id"
    >
      <span
        class="euiButtonContent euiButton__content"
      >
        <span
          class="euiButton__text euiButtonGroupButton__textShift"
          data-text="Visual Editor"
          title="Visual Editor"
        >
          <input
            checked=""
            class="euiScreenReaderOnly"
            data-test-subj="VISUAL"
            id="some_html_id"
            name="some_html_id"
            type="radio"
            value=""
          />
          Visual Editor
        </span>
      </span>
    </label>
    <label
      class="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small"
      data-test-subj="editor-type-json-editor"
      for="some_html_id"
    >
      <span
        class="euiButtonContent euiButton__content"
      >
        <span
          class="euiButton__text euiButtonGroupButton__textShift"
          data-text="JSON Editor"
          title="JSON Editor"
        >
          <input
            class="euiScreenReaderOnly"
            data-test-subj="JSON"
            id="some_html_id"
            name="some_html_id"
            type="radio"
            value=""
          />
          JSON Editor
        </span>
      </span>
    </label>
  </div>
</fieldset>
`;
